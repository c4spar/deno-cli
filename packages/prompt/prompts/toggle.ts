import { blue, dim, underline } from 'https://deno.land/std@v0.61.0/fmt/colors.ts';
import { KeyEvent } from '../../keycode/lib/key-event.ts';
import { Figures } from '../lib/figures.ts';
import { GenericPrompt, GenericPromptOptions, GenericPromptSettings } from '../lib/generic-prompt.ts';

export interface ToggleOptions extends GenericPromptOptions<boolean, string> {
    active?: string;
    inactive?: string;
}

export interface ToggleSettings extends GenericPromptSettings<boolean, string> {
    active: string;
    inactive: string;
}

export class Toggle extends GenericPrompt<boolean, string, ToggleSettings> {

    protected status: string = typeof this.settings.default !== 'undefined' ? this.format( this.settings.default ) : '';

    public static async prompt( options: string | ToggleOptions ): Promise<boolean> {

        if ( typeof options === 'string' ) {
            options = { message: options };
        }

        return new this( {
            pointer: blue( Figures.POINTER_SMALL ),
            active: 'Yes',
            inactive: 'No',
            ...options
        } ).prompt();
    }

    protected setPrompt( message: string ) {

        message += ` ${ this.settings.pointer } `;

        if ( this.status === this.settings.active ) {
            message += `${ dim( `${ this.settings.inactive } /` ) } ${ underline( this.settings.active ) }`;
        } else if ( this.status === this.settings.inactive ) {
            message += `${ underline( this.settings.inactive ) } ${ dim( `/ ${ this.settings.active }` ) }`;
        } else {
            message += dim( `${ this.settings.inactive } / ${ this.settings.active }` );
        }

        this.write( message );
    }

    protected async read(): Promise<boolean> {

        this.screen.cursorHide();

        return super.read();
    }

    protected async handleEvent( event: KeyEvent ): Promise<boolean> {

        switch ( event.name ) {

            case 'c':
                if ( event.ctrl ) {
                    return Deno.exit( 0 );
                }
                break;

            case this.settings.inactive[ 0 ].toLowerCase():
            case 'n': // no nein non
            case 'left':
                this.selectInactive();
                break;

            case this.settings.active[ 0 ].toLowerCase():
            case 'y': // yes
            case 'j': // ja
            case 's': // si
            case 'o': // oui
            case 'right':
                this.selectActive();
                break;

            case 'return':
            case 'enter':
                return true;
        }

        return false;
    }

    protected selectActive() {
        this.status = this.settings.active;
    }

    protected selectInactive() {
        this.status = this.settings.inactive;
    }

    protected validate( value: string ): boolean | string {
        return [ this.settings.active, this.settings.inactive ].indexOf( value ) !== -1;
    }

    protected transform( value: string ): boolean | undefined {
        switch ( value ) {
            case this.settings.active:
                return true;
            case this.settings.inactive:
                return false;
        }
        return;
    }

    protected format( value: boolean ): string {
        return value ? this.settings.active : this.settings.inactive;
    }

    protected getValue(): string {
        return this.status;
    }
}
