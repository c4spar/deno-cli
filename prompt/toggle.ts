import type { KeyEvent } from "../keycode/key_event.ts";
import { blue, dim, underline } from "./deps.ts";
import { Figures } from "./figures.ts";
import {
  GenericPrompt,
  GenericPromptOptions,
  GenericPromptSettings,
} from "./_generic_prompt.ts";

/** Toggle key options. */
export interface ToggleKeys {
  active?: string[];
  inactive?: string[];
  submit?: string[];
}

/** Toggle key settings. */
type ToggleKeysSettings = Required<ToggleKeys>;

/** Generic prompt options. */
export interface ToggleOptions extends GenericPromptOptions<boolean, string> {
  active?: string;
  inactive?: string;
  keys?: ToggleKeys;
}

/** Toggle prompt settings. */
interface ToggleSettings extends GenericPromptSettings<boolean, string> {
  active: string;
  inactive: string;
  keys: ToggleKeysSettings;
}

/** Toggle prompt representation. */
export class Toggle extends GenericPrompt<boolean, string, ToggleSettings> {
  protected status: string = typeof this.settings.default !== "undefined"
    ? this.format(this.settings.default)
    : "";

  /** Execute the prompt and show cursor on end. */
  public static prompt(
    options: string | ToggleOptions,
  ): Promise<boolean> {
    if (typeof options === "string") {
      options = { message: options };
    }

    return new this({
      pointer: blue(Figures.POINTER_SMALL),
      active: "Yes",
      inactive: "No",
      ...options,
      keys: {
        active: ["right", "y", "j", "s", "o"],
        inactive: ["left", "n"],
        submit: ["enter", "return"],
        ...(options.keys ?? {}),
      },
    }).prompt();
  }

  /**
   * Set prompt message.
   * @param message Prompt message.
   */
  protected render(message: string) {
    message += " " + this.settings.pointer + " ";

    if (this.status === this.settings.active) {
      message += dim(this.settings.inactive + " /") + " " +
        underline(this.settings.active);
    } else if (this.status === this.settings.inactive) {
      message += underline(this.settings.inactive) + " " +
        dim("/ " + this.settings.active);
    } else {
      message += dim(this.settings.inactive + " / " + this.settings.active);
    }

    this.clear();
    this.write(message);
  }

  /** Read user input from stdin, handle events and validate user input. */
  protected read(): Promise<boolean> {
    this.screen.cursorHide();
    return super.read();
  }

  /**
   * Handle user input event.
   * @param event Key event.
   */
  protected handleEvent(event: KeyEvent): boolean {
    switch (true) {
      case event.name === "c":
        if (event.ctrl) {
          this.screen.cursorShow();
          return Deno.exit(0);
        }
        break;

      case event.sequence === this.settings.inactive[0].toLowerCase():
      case this.isKey(this.settings.keys, "inactive", event):
        this.selectInactive();
        break;

      case event.sequence === this.settings.active[0].toLowerCase():
      case this.isKey(this.settings.keys, "active", event):
        this.selectActive();
        break;

      case this.isKey(this.settings.keys, "submit", event):
        return true;
    }

    return false;
  }

  /** Set active. */
  protected selectActive() {
    this.status = this.settings.active;
  }

  /** Set inactive. */
  protected selectInactive() {
    this.status = this.settings.inactive;
  }

  /**
   * Validate input value.
   * @param value User input value.
   * @return True on success, false or error message on error.
   */
  protected validate(value: string): boolean | string {
    return [this.settings.active, this.settings.inactive].indexOf(value) !== -1;
  }

  /**
   * Map input value to output value.
   * @param value Input value.
   * @return Output value.
   */
  protected transform(value: string): boolean | undefined {
    switch (value) {
      case this.settings.active:
        return true;
      case this.settings.inactive:
        return false;
    }
  }

  /**
   * Format output value.
   * @param value Output value.
   */
  protected format(value: boolean): string {
    return value ? this.settings.active : this.settings.inactive;
  }

  /** Get input value. */
  protected getValue(): string {
    return this.status;
  }
}
