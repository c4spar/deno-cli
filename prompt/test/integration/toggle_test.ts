import { Toggle } from "../../toggle.ts";
import { assertSnapshotCall } from "../../../testing/assert_snapshot_call.ts";

await assertSnapshotCall({
  name: "toggle prompt",
  meta: import.meta,
  osSuffix: ["windows"],
  steps: {
    "should toggle the prompt": { stdin: ["y", "n", "y", "\n"] },
  },
  async fn() {
    await Toggle.prompt({
      message: "Please confirm",
      hint: "some hint",
      default: false,
    });
  },
});
