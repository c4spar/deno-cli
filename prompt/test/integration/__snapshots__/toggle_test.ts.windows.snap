export const snapshot = {};

snapshot[`assert prompt snapshot - toggle_test.ts 1`] = `
"stdout:
? Please confirm (No) » No / Yes
❯ some hint\\x1b[1A\\x1b[0G\\x1b[?25l\\x1b[G\\x1b[0J
? Please confirm (No) » No / Yes
❯ some hint\\x1b[1A\\x1b[0G\\x1b[?25l\\x1b[G\\x1b[0J
? Please confirm (No) » No / Yes
❯ some hint\\x1b[1A\\x1b[0G\\x1b[?25l\\x1b[G\\x1b[0J
? Please confirm (No) » No / Yes
❯ some hint\\x1b[1A\\x1b[0G\\x1b[?25l\\x1b[G\\x1b[0J
? Please confirm (No) » Yes
\\x1b[?25h\\x1b[?25h
stderr:
"
`;
